ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java

```diff
@@ -600,36 +622,61 @@ protected void deleteBlob(final String blobIdStr, final String serviceId, Future
+          if (exception != null) {
+            completeOperation(futureResult, callback, null, exception, false);
+          } else if (result.getBlobResult != null) {
+            exception = new RouterException(
+                "GET blob call returned the blob instead of just the store keys (before TTL update)",
+                RouterErrorCode.UnexpectedInternalError);
+            completeOperation(futureResult, callback, null, exception, false);
+          } else {
+            List<String> blobIdStrs = new ArrayList<>();
+            blobIdStrs.add(blobIdStr);
+            if (result.storeKeys != null) {
+              result.storeKeys.forEach(key -> blobIdStrs.add(key.getID()));
+            }
+            // TTLUpdateManager updates all chunks in parallel
+            currentOperationsCount.addAndGet(blobIdStrs.size());
+            doUpdateTtlOperation(blobIdStrs, serviceId, expiresAtMs, futureResult, callback);
           }
+        };
+
+        GetBlobOptionsInternal options =
+            new GetBlobOptionsInternal(new GetBlobOptions(GetBlobOptions.OperationType.All, GetOption.None, null), true,
+                routerMetrics.ageAtTtlUpdate);
+        try {
+          getBlob(blobIdStr, options, internalCallback);
+        } catch (RouterException e) {
+          completeUpdateBlobTtlOperation(e, futureResult, callback);
         }
-      };
-      GetBlobOptionsInternal options =
-          new GetBlobOptionsInternal(new GetBlobOptions(GetBlobOptions.OperationType.All, GetOption.None, null), true,
-              routerMetrics.ageAtTtlUpdate);
+      } else {
+        // needGet is false => do update directly on single blobId
+        routerMetrics.skippedGetBlobCount.inc();
+        doUpdateTtlOperation(Collections.singletonList(blobIdStr), serviceId, expiresAtMs, futureResult, callback);
+      }
+    }
+
+    /**
+     * Helper method that submits the TTL update operation and handles exceptions.
+     * @param blobIdStrs The original blobId strings
+     * @param serviceId The service ID of the service updating the ttl of the blob(s). This can be null if unknown.
+     * @param expiresAtMs The new expiry time (in ms) of the blob.
+     * @param futureResult The {@link FutureResult} that will contain the result eventually and exception if any.
+     * @param callback The {@link Callback} that will be called on completion of the request.
+     */
+    private void doUpdateTtlOperation(List<String> blobIdStrs, final String serviceId, long expiresAtMs,
+        FutureResult<Void> futureResult, Callback<Void> callback) {
       try {
-        getBlob(blobIdStr, options, internalCallback);
+        ttlUpdateManager.submitTtlUpdateOperation(blobIdStrs, serviceId, expiresAtMs, futureResult, callback);
+        routerCallback.onPollReady();
       } catch (RouterException e) {
+        currentOperationsCount.addAndGet(1 - blobIdStrs.size());
```

Comment:
how is this `1 - blobIdStrs.size()`? I think a comment would be useful. Also what happens in the successful case, where does this count get decremented? Where does `completeOperation()` get called basically?

Reply:
This line was identical before my code change.  I agree an explanation comment would be helpful, but doubt that I'm best qualified to add it.